### PR TITLE
AirfRANS: gc=1.5 + T_max=5 (no WD) — isolate T_max=5 at new winning clip

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Grad-clip=1.5 is the new AirfRANS winner (PR #2737: 0.00935). Separately, T_max=5 improved 0.01323→0.01271 at gc=1.0 (PR #2732). These two improvements have not been combined in isolation yet.

This experiment tests gc=1.5 + T_max=5 **without weight decay** to isolate the T_max effect at the new winning clip value. PR #2743 (haku) tests gc=1.5+T_max=5+WD=1e-2 — this PR complements that by removing WD entirely, making it possible to attribute any delta to T_max=5 vs WD independently.

With gc=1.5, the LR spike suppression is stronger — cosine peaks no longer cause destructive gradient explosions. In that regime, shorter cycles (T_max=5 → 8 restarts per 40 epochs) may compound the basin-exploration benefit further, since each cosine trough is a clean landing opportunity without spikes derailing the trajectory.

Expected outcome: T_max=5 at gc=1.5 should find a deeper basin than T_max=10 at gc=1.5, potentially pushing well below 0.009.

## Instructions

Run the following training command exactly as written:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 1.5 \
  --no-use-ema \
  --enable-fourier \
  --epochs 999 \
  --wandb_group airfrans-gc15-tmax5
```

Key differences from baseline (#2737):
- `--cosine-t-max 5` (baseline uses T_max=10)
- `--grad-clip 1.5` (same as baseline winner)
- No `--weight-decay` flag (default WD, isolating T_max effect)

Report:
1. Best `val_primary/surface_mse` and at which epoch
2. Whether best is at a cosine trough (at T_max=5, troughs fall at epochs 5, 10, 15, 20, 25, 30, 35, 40)
3. Any gradient spike observations compared to baseline
4. W&B run ID

## Baseline

- **Dataset:** AirfRANS (task: full)
- **Primary metric:** `val_primary/surface_mse` (lower is better)
- **Current best:** 0.00935 — PR #2737 (haku, gc=1.5, T_max=10, lr=7e-4, no WD)
- **Previous T_max=5 result:** 0.01271 — PR #2732 (at gc=1.0, WD=1e-4)
- **Baseline W&B run:** 7bdiqnmi
- **External target:** 0.0043 (~2.2x gap remaining)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --grad-clip 1.5 --no-use-ema --enable-fourier --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kohaku:

## Results

### Key Metrics

| Metric | This Run (gc=1.5, T_max=5) | Baseline (gc=1.5, T_max=10) | Delta |
|---|---|---|---|
| **val_primary/surface_mse** | **0.00978** (epoch 43) | **0.00935** (epoch 40) | **+4.6% (worse)** |
| full_val/surface_mse_p | 0.03898 | — | — |
| full_val/volume_mse | 0.06795 | — | — |
| test_primary/surface_mse | 0.01990 (final epoch) | — | — |

### Training Command

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.5 \
  --no-use-ema --enable-fourier --epochs 999 \
  --wandb-group airfrans-gc15-tmax5 \
  --agent kohaku --wandb-name "kohaku/airfrans-gc15-tmax5-noWD"
```

### Run Details

- **W&B run ID:** 54g6lnx2
- **Epochs:** 50 (hit SENPAI_MAX_EPOCHS cap, not timeout)
- **Runtime:** ~26 minutes (1556s)
- **Peak memory:** ~98GB VRAM (Blackwell RTX PRO 6000)
- **GPU:** NVIDIA RTX PRO 6000 Blackwell Server Edition

### Cosine Trough Analysis

With T_max=5, troughs fall at epochs 5, 10, 15, 20, 25, 30, 35, 40, 45, 50.

**Best was at epoch 43 — NOT at a trough.** This is unusual. The nearest troughs are epoch 40 (0.0897) and epoch 45 (0.0170). The model briefly found a deep basin mid-cycle at epoch 43 but was kicked out by rising LR. Trough values were highly volatile:

| Trough Epoch | val_primary/surface_mse |
|---|---|
| 5 | 0.25336 |
| 10 | 0.18354 |
| 15 | 0.12447 |
| 20 | 0.06577 |
| 25 | 0.03373 |
| 30 | 0.02276 |
| 35 | 0.04590 |
| 40 | 0.08967 |
| 45 | 0.01699 |
| 50 | 0.01627 |

The trough envelope is non-monotonic (troughs 35 and 40 spike up), unlike T_max=10's more consistent descent to 0.00935.

### Gradient Observations

- Grad clip events: 254–348 per epoch (70–97% of 360 batches clipped)
- Grad norm mean: ~11–17 early → ~10–13 later (consistent with baseline)
- No catastrophic gradient spikes observed — gc=1.5 effectively suppresses them at T_max=5 as it does at T_max=10

### What Happened

**T_max=5 at gc=1.5 did NOT beat the T_max=10 baseline.** The result (0.00978) is close but ~4.6% worse.

The hypothesis was that shorter cycles (8 restarts per 40 epochs vs 4 for T_max=10) would compound the basin-exploration benefit at gc=1.5. Instead, the shorter cycles create too much volatility — the model doesn't have enough gradient steps per cycle to descend deeply into basins before LR ramps back up. With only 5 epochs per cycle (5 × 360 = 1800 gradient steps), the optimization trajectory is too erratic to settle compared to T_max=10's 10 epochs (3600 steps) per cycle.

The non-monotonic trough envelope (troughs at 35 and 40 spiking to 0.046 and 0.090) confirms the model is being kicked out of basins faster than it can re-converge. The best value occurring mid-cycle (epoch 43, not at a trough) further suggests the model lands in transient states that don't survive the next LR peak.

**Key insight:** At gc=1.5, T_max=10 remains the optimal cycle length. The gc=1.5 clipping suppresses spikes but doesn't compensate for the reduced per-cycle convergence of shorter restarts.

### Suggested Follow-ups

1. **T_max=15 or T_max=20 at gc=1.5** — if T_max=5 is too short and T_max=10 is optimal, try slightly longer cycles to see if the basin is even deeper with more steps per cycle
2. **T_max=5 with longer training** — the trough envelope at epochs 45/50 (0.017/0.016) is still descending; with more epochs, T_max=5 might eventually catch up as later troughs compound
3. **Lower lr with T_max=5** — the high volatility might be addressed by reducing lr from 7e-4 to 5e-4, giving the shorter cycles a gentler optimization landscape

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2748 because ADVISOR: Closing — gc=1.5+T_max=5 achieved 0.009778, doesn't beat 0.007264 baseline. T_max=5 is too volatile with gc=1.5. 4L/256d architecture is the path forward..
